include/tvm/relay/module.h

```diff
@@ -43,11 +43,18 @@ class ModuleNode : public RelayNode {
   /*! \brief A map from ids to all global functions. */
   tvm::Map<GlobalVar, Function> functions;
 
+  /*! \brief The error reporter. */
+  mutable ErrorReporter err_reporter;
```

Comment:
It is a bit strange that error reporter is a field of the module. Module itself should only contain data, but not the procedure or auxiliary data structure to support the module, I would rather introduce something like TLS(thread local state) for error reporting and context

Reply:
In my view the module is the global data about a set of functions, including the errors. Errors only make sense in the context of a single module, and have no bearing on the validity of other modules (i.e I believe errors should be local to a module).

Putting them in a thread local global variable lifts the level of errors from the module to the thread. Processing modules on the same thread should not intermingle their errors.

In my experience it is better have self-contained objects, otherwise functionality becomes dependent on the context in which the object is being used which can lead to frustrating bugs for the end-user.

Reply:
My main concern about coupling error reporter (as a mutable state to the module) is that it complicates the reasoning of the user. If we have a way to make it immutable, and only contains the objects(but not necessarily the methods for error reporting) it might be fine. But this will imply that error is carried over within a module.

Reply:
To summarize my comment, i think we need to separate the data that persists between pass(meta data to help error reporting) v.s. the function logic to do the reporting action